File: duna-core/src/main/java/io/duna/core/internal/eventbus/LocalEventBus.java
package io.duna.core.internal.eventbus;

import io.duna.core.concurrent.future.Future;
import io.duna.core.eventbus.EventBus;
import io.duna.core.eventbus.EventRouter;
import io.duna.core.eventbus.Message;
import io.duna.core.eventbus.event.Event;
import io.duna.core.eventbus.event.InboundEvent;
import io.duna.core.eventbus.event.OutboundEvent;
import io.duna.core.internal.eventbus.event.DefaultInboundEvent;
import io.duna.core.internal.eventbus.event.DefaultOutboundEvent;

import java.util.UUID;
import java.util.function.Consumer;

public class LocalEventBus implements EventBus {

    private EventRouter router;

    public LocalEventBus() {
        this.router = new LocalEventRouter();
    }

    @Override
    public <T> OutboundEvent<T> outbound(String name) {
        return new DefaultOutboundEvent<T>(this, name);
    }

    @Override
    public <T> OutboundEvent<T> outbound(String name, int cost) {
        return new DefaultOutboundEvent<T>(this, name).setCost(cost);
    }

    @Override
    public <T> InboundEvent<T> inbound() {
        String name = UUID.randomUUID().toString();
        return new DefaultInboundEvent<>(this, name);
    }

    @Override
    public <T> InboundEvent<T> inbound(String name) {
        return new DefaultInboundEvent<>(this, name);
    }

    @Override
    public <T> InboundEvent<T> inbound(String name, int cost) {
        return new DefaultInboundEvent<T>(this, name).setCost(cost);
    }

    @Override
    public <T> Future<Void> dispatch(Message<T> outgoing, Consumer<Message<T>> deliveryErrorConsumer) {
        process(outgoing); // Dispatches to the local eventbus
        return Future.completedFuture();
    }

    @Override
    public <T> Future<Void> publish(Message<T> outgoing, Consumer<Message<T>> deliveryErrorConsumer) {

        /* The implementation here is the same as `dispatch` because you can only have
         * one event registered per eventbus node. In distributed environments,
         * this should broadcast the message to all nodes in the cluster.
         */
        process(outgoing);
        return Future.completedFuture();
    }

    @Override
    public <T> void process(Message<T> incoming) {
        // Process the event in the thread corresponding to the event
        Event<?> inbound = router.get(incoming.getTarget());

        if (inbound == null) {
            // Log and exit
            return;
        }

        if (inbound instanceof DefaultInboundEvent) {
            // Should accept in another event loop
            ((DefaultInboundEvent<T>) inbound).accept(incoming);
        }
    }

    public boolean register(Event<?> event) {
        if (router.contains(event.getName()))
            return false;

        router.register(event);
        return true;
    }

    public void cancel(Event<?> event) {
        router.cancel(event);
    }

    public <T> T poll(String queue, Event<?> event) {
        return null;
    }
}

File: duna-core/src/main/kotlin/io/duna/core/eventbus/impl/LocalEventBus.kt
package io.duna.core.eventbus.impl

import io.duna.core.concurrent.Future
import io.duna.core.eventbus.EventBus
import io.duna.core.eventbus.Message
import io.duna.core.eventbus.event.DefaultEvent
import io.duna.core.eventbus.event.DefaultSubscriber
import io.duna.core.eventbus.event.Event
import io.duna.core.eventbus.event.Subscriber
import io.duna.core.eventbus.queue.MessageQueue
import io.duna.core.eventbus.routing.LocalRouter
import io.duna.core.eventbus.routing.Router
import io.netty.channel.EventLoop
import io.netty.channel.EventLoopGroup
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.ExecutorService

internal class LocalEventBus(private val eventLoopGroup: EventLoopGroup,
                             private val workerPool: ExecutorService) : EventBus {

  private val pipeline = PipelineImpl()
  private val router = LocalRouter()
  private val messageQueues = ConcurrentHashMap<String, MessageQueue<*>>()

  private val eventExecutorMap = ConcurrentHashMap<Subscriber<*>, EventLoop>()

  override fun <T> emit(name: String): Event<T> = DefaultEvent(this, name)

  override fun <T> subscriber(event: String): Subscriber<T> {
    val subscriber = DefaultSubscriber<T>(this, event)
    eventExecutorMap[subscriber] = eventLoopGroup.next()

    return subscriber
  }

  @Suppress("UNCHECKED_CAST")
  override fun <T> queue(name: String): MessageQueue<T> = messageQueues.computeIfAbsent(name) {
    TODO()
  } as MessageQueue<T>

  override fun router(): Router = router

  override fun accept(message: Message<*>) {

  }

  override fun dispatch(message: Message<*>): Future<Message<*>> {
    accept(message)
    return Future.completedFuture()
  }

  override fun purge(subscriber: Subscriber<*>): Boolean {
    TODO("not implemented")
  }
}
